RecoBTag/FeatureTools/plugins/ParticleNetFeatureEvaluator.cc

```diff
@@ -0,0 +1,807 @@
+        std::abs(jet.eta()) > max_jet_eta_ or std::abs(jet.eta()) < min_jet_eta_)
+      fill_vars = false;
+    if (jet.numberOfDaughters() == 0)
+      fill_vars = false;
+
+    // fill features
+    if (fill_vars) {
+      fillParticleFeatures(features, jet, tau_pfcandidates, *muons, *electrons, *photons);
+      fillSVFeatures(features, jet);
+      fillLostTrackFeatures(features, jet);
+      features.check_consistency(particle_features_);
+      features.check_consistency(sv_features_);
+      features.check_consistency(losttrack_features_);
+    }
+
+    // this should always be done even if features are not filled
+    output_tag_infos->emplace_back(features, jet_ref);
+  }
+  // move output collection
+  iEvent.put(std::move(output_tag_infos));
+}
+
+bool ParticleNetFeatureEvaluator::useTrackProperties(const pat::PackedCandidate *cand) {
+  const auto *track = cand->bestTrack();
+  return track != nullptr and track->pt() > min_pt_for_track_properties_;
+};
+
+void ParticleNetFeatureEvaluator::fillParticleFeatures(DeepBoostedJetFeatures &fts,
+                                                       const reco::Jet &jet,
+                                                       const std::vector<math::XYZTLorentzVector> &tau_pfcandidates,
+                                                       const pat::MuonCollection &muons,
+                                                       const pat::ElectronCollection &electrons,
+                                                       const pat::PhotonCollection &photons) {
+  // some jet properties
+  math::XYZVector jet_dir = jet.momentum().Unit();
+  TVector3 jet_direction(jet.momentum().Unit().x(), jet.momentum().Unit().y(), jet.momentum().Unit().z());
+  GlobalVector jet_ref_track_dir(jet.px(), jet.py(), jet.pz());
+  // vertexes
+  reco::VertexRefProd PVRefProd(vtxs_);
+  // track builder
+  TrackInfoBuilder trackinfo(track_builder_);
+
+  // make list of pf-candidates to be considered
+  std::vector<const pat::PackedCandidate *> daughters;
+  for (const auto &dau : jet.daughterPtrVector()) {
+    // remove particles w/ extremely low puppi weights
+    const pat::PackedCandidate *cand = dynamic_cast<const pat::PackedCandidate *>(&(*dau));
+    if (not cand)
+      throw edm::Exception(edm::errors::InvalidReference) << "Cannot convert to either pat::PackedCandidate";
+    if (cand->puppiWeight() < min_puppi_wgt_)
+      continue;
+    // base requirements on PF candidates
+    if (cand->pt() < min_pt_for_pfcandidates_)
+      continue;
+    // charged candidate selection (for Higgs Interaction Net)
+    if (!include_neutrals_ and (cand->charge() == 0 or cand->pt() < min_pt_for_track_properties_))
+      continue;
+    // filling daughters
+    daughters.push_back(cand);
+  }
+
+  // sort by Puppi-weighted pt
+  if (use_puppiP4_)
+    std::sort(daughters.begin(), daughters.end(), [&](const pat::PackedCandidate *a, const pat::PackedCandidate *b) {
```

Comment:
As far as I can make out, `use_puppiP4` is disabled by default and not enabled for PUPPI jets. Is that expected? When training on PUPPI jets, don't you use the PUPPI weights when defining the input features?

Also, there has been some ongoing work from JME (@nurfikri89 ) to retrieve the PUPPI weights from ValueMaps instead of from the candidates, in case the weights are re-computed in a re-mini or re-nano, see for instance: https://github.com/cms-sw/cmssw/pull/40803
Shouldn't this also be possible here?

Reply:
We are not using the PUPPI weight to define the pT of the PF candidate constituents of the jet, but rather directly the PF candidate pT. However, we are only using the PF candidates associated with the PUPPI jet, meaning that all PF candidates pass a minimum threshold on the PUPPI weight. We could have also weighted the PF candidate pT by the PUPPI weight. We did not explicitly try this, but we did perform a training in which we added the PUPPI weight as an additional PF candidate input feature and saw no noticeable difference in performance. We therefore chose to omit the PUPPI weight from the input features for the sake of minimizing the dependence of the ParticleNet training on the PUPPI version used. The use_puppiP4 boolean is inherited from what is done in the DeepBoostedJetTagInfoProducer, and could be removed from the ParticleNetFeatureEvaluator. How about we remove it for the time being, with the understanding that if we change the training to include the PUPPI-weighted features we just port the changes from the other PR?

Reply:
Discussed this further with @swertz privately, we should be on the same page now. Agreed to just remove the PUPPI weight-related parts of the ParticleNetFeatureEvaluator, since they are anyway not used.

Reply:
Hi, sorry to come back to this (maybe better to discuss it offline also with JMAR).
Just wanted to mention that the training will depend in any case on the puppi version:
1) I remember seeing an effect on the tau tagging deepening on the puppi version
2) you're using jets that are clustered with Puppi weights, so they are substantially different compared to CHS or even PF jets.

I think that for one we can go ahead, but sooner rather than later we should re-think this decision on whether or not we want to use puppi weights for consistency.

Reply:
Hi @anmalara, yes sure the training will always depend on the PUPPI version since the PF candidate inputs for the jet will only be those associated to the jet with PUPPI weight above a minimum threshold. That being said, we did not include the PUPPI weights in the training directly because we saw that it did not improve the performance and because this could in principle increase the dependency of the training on the PUPPI version. We evaluated a recent CHS jet training on PUPPI jets and saw that the performance degraded noticeably, at the level of a few percent in b efficiency for a given mistag. That being said, this is a change that likely goes well beyond what is expected from changes in PUPPI versions, so it's probably a very conservative upper limit on the susceptibility of the network performance to PUPPI tune. Then about the taus - yes the effect was not really the tau tagging performance itself but rather that PUPPI jets are inefficient seeds for hadronic taus. In other words, a significant fraction of hadronic tau decay PF candidates can get a sufficiently small PUPPI weight to not be included in the PF constituents of the jet in the first place. The changes in this PR are more meant to enable usage of these networks for q and g jet reconstruction. The integration of these networks within the tau reconstruction is an ongoing work by @mbluj .

Reply:
I discussed this further with @rgerosa - we will try an alternative training now where we give the PF candidate pT input variable as puppi-weighted and compare with the performance of the training in this PR, which has no PUPPI weight in the PF cand. features. That being said, we think it's fine to go ahead with integration of these networks/PR in the meantime.